dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java

```diff
@@ -34,33 +36,119 @@
+   * @throws IOException when there is a problem in reading a jar file
+   * @throws ClassNotFoundException when there is a problem in reading a class from a jar file
+   */
+  public static void main(String[] arguments) throws IOException, ClassNotFoundException {
+    String report = generateStaticLinkageReport(arguments);
+    System.out.println(report);
+  }
+
+  @VisibleForTesting
+  static String generateStaticLinkageReport(String[] jarFileNames) throws IOException,
+      ClassNotFoundException {
+    StringBuilder stringBuilder = new StringBuilder();
+    Set<FullyQualifiedMethodSignature> externalMethodReferences = new HashSet<>();
+    List<Path> paths = new ArrayList<>();
+    for (String jarFileName : jarFileNames) {
+      File jarFile = new File(jarFileName);
+      Path jarFilePath = jarFile.toPath();
+      paths.add(jarFilePath);
+      externalMethodReferences.addAll(listExternalMethodReferences(jarFilePath));
+    }
+    List<FullyQualifiedMethodSignature> unresolvedMethodReferences =
+        resolvedMethodReferences(paths, Lists.newArrayList(externalMethodReferences));
+    if (unresolvedMethodReferences.isEmpty()) {
+      stringBuilder.append("There were no unresolved method references from the jar file(s) :");
+      stringBuilder.append(paths);
+    } else {
+      stringBuilder.append("There were unresolved method references from the jar file(s):\n");
+      for (FullyQualifiedMethodSignature methodReference : unresolvedMethodReferences) {
+        stringBuilder.append("  ");
+        stringBuilder.append(methodReference);
+        stringBuilder.append("\n");
+      }
+    }
+    return stringBuilder.toString();
+  }
+
+  /**
+   * Checks the availability of the methods through the jar files, and lists the unavailable
+   * methods.
+   *
+   * @param paths jar files to search for the availability of the methods
+   * @param methodReferences methods to search for with the jar files
+   * @return list of methods that are not found in the jar files
+   */
+  static List<FullyQualifiedMethodSignature> resolvedMethodReferences(List<Path> paths,
+      List<FullyQualifiedMethodSignature> methodReferences) {
+    List<FullyQualifiedMethodSignature> unresolvedMethods = new ArrayList<>();
+
+    // Creates chain of ClassPath items in the same order as paths
+    ClassPath classPath = null;
+    for (Path path : paths) {
+      String pathFileName = path.toFile().getAbsolutePath();
+      classPath = new ClassPath(classPath, pathFileName);
+    }
+    SyntheticRepository repository = SyntheticRepository.getInstance(classPath);
+
+    for (FullyQualifiedMethodSignature methodReference : methodReferences) {
+      try {
+        JavaClass javaClass = repository.loadClass(methodReference.getClassName());
+        MethodSignature methodSignature = methodReference.getMethodSignature();
+        boolean methodFound = false;
+        // Opportunity to get better performance to create a cache (from class name to method list)
```

Comment:
I don't think method list is what you need - you need a Set that you can use to look up method references in O(1) time instead of O(n) time.

Reply:
Let me add cache for this.

Reply:
Done.

Reply:
Now it uses set to check the methods that's already available:

https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/121/files#diff-cb2eefbf6e8ca674c445c1c7e1b519edR133